code/datums/progress_bar.dm

```diff
@@ -0,0 +1,182 @@
+	var/client/client
+	/// The bar's index number within the owning mob's active progress bar list.
+	var/list_index
+	/// Whether the bar is stopping.
+	var/stopping = FALSE
+
+/datum/progress_bar/New(mob/user, goal, atom/target)
+	. = ..()
+	if(!istype(target))
+		EXCEPTION("Invalid target given")
+
+	init_images(target)
+	src.target = target
+	src.user = user
+	src.goal = goal
+
+	if(user)
+		client = user.client
+		client?.images |= holder
+
+		LAZYINITLIST(user.progressbars)
+		LAZYINITLIST(user.progressbars[target])
+		var/list/bars = user.progressbars[target]
+		bars += src
+		list_index = length(bars)
+		animate(holder, pixel_z = world.icon_size + (height * (list_index - 1)), alpha = 255, time = 0.5 SECONDS, easing = SINE_EASING | EASE_OUT)
+	else
+		animate(holder, alpha = 255, time = 0.5 SECONDS, easing = SINE_EASING | EASE_OUT)
+
+/datum/progress_bar/Destroy()
+	client?.images -= holder
+	QDEL_NULL(holder)
+	QDEL_NULL(backdrop)
+	QDEL_NULL(fill)
+	return ..()
+
+/**
+  * Stops the progress bar and removes it after a bit.
+  */
+/datum/progress_bar/proc/stop()
+	if(stopping)
+		return
+	if(user?.progressbars[target])
+		var/list/bars = user.progressbars[target]
+		for(var/I in (bars - src))
+			var/datum/progress_bar/P = I
+			if(P.list_index > list_index)
+				P.shift_down()
+
+		bars -= src
+		if(!length(bars))
+			LAZYREMOVE(user.progressbars, target)
+
+	stopping = TRUE
+	animate(holder, alpha = 0, time = 0.5 SECONDS)
+	QDEL_IN(src, 0.5 SECONDS)
+
+/**
+  * Initializes the visual elements.
+  *
+  * Arguments:
+  * * target - The atom to parent the images to.
+  */
+/datum/progress_bar/proc/init_images(target)
+	holder = image('icons/effects/effects.dmi', target, "nothing", HUD_ABOVE_ITEM_LAYER)
+	holder.mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	holder.alpha = 0
+	holder.plane = HUD_PLANE
+	holder.appearance_flags = KEEP_TOGETHER | APPEARANCE_UI_IGNORE_ALPHA
+
+	backdrop = image(icon_path, target, "[theme]_backdrop")
+
+	fill = image(icon_path, target, "[theme]_fill")
+	fill.filters = filter(type = "alpha", icon = icon(icon_path, "[theme]_mask"), x = -fill_width)
+
+	holder.overlays += backdrop
+	holder.overlays += fill
+
+/**
+  * Call periodically to update the progress bar's progress.
+  *
+  * Arguments:
+  * * progress - The new value, divided with the given goal on creation to calculate the fraction.
+  */
+/datum/progress_bar/proc/update(progress)
+	if(QDELETED(target))
+		qdel(src)
+		return
+	if(user?.client != client)
+		client?.images -= holder
+		if(user.client)
+			client = user.client
+			client.images += holder
+
+	. = clamp(progress / goal, 0, 1)
+	if(fill)
+		holder.overlays -= fill
+		update_fill(.)
+		holder.overlays += fill
+
+/**
+  * Updates the fill. Do your visual changes here.
+  *
+  * Arguments:
+  * * fraction - The new (value / goal) fraction.
+  */
+/datum/progress_bar/proc/update_fill(fraction)
+	fill?.filters[1]?:x = -fill_width * (1 - fraction)
```

Comment:
Use `?.`, or break out the code if need be.

Reply:
/datum/progress_bar/proc/update_fill works as is because messing with filters is painful, and we dealing with datum which don't support filter helpers. Changing the look will produce more lines or even make dreamchecker scream.

Reply:
I'm fine with more lines. If there are helpers to avoid this that don't work, improve the helpers. The only excuse for doing it like this is that it's literally impossible to write functionally equivalent code in a different way.

Reply:
Replaced with `?.`, comment outdated.